libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+namespace
+{
+
+vector<Declaration const*> allAnnotatedDeclarations(Identifier const* _identifier)
+{
+	vector<Declaration const*> output;
+	output.push_back(_identifier->annotation().referencedDeclaration);
+	output += _identifier->annotation().candidateDeclarations;
+	return output;
+}
+
+}
+
+ReferenceCollector::ReferenceCollector(
+	Declaration const& _declaration,
+	std::string const& _sourceIdentifierName
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _ast,
+	std::string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_ast.accept(collector);
+	return std::move(collector.m_resultingReferences);
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto output = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);
+			break;
+		}
```

Comment:
Took me a moment to understand that it's just a simple nested `if`. Wouldn't it be simpler to do this? 
```suggestion
		if (m_sourceIdentifierName == *symbolAlias.alias)
		{
			if (symbolAlias.symbol && symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration)
				m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
			else
				m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);

			break;
		}
```